### PR TITLE
feat: adding polymorphic relations support for models

### PR DESCRIPTION
**Pull request by OnTheThirdDay**
Addressing the feature parity of polymorphic relations from lb3, and the syntax is similar to lb3. The changes are backward compatible with the previous APIs. hasOne, belongsTo, and hasManyThrough relations are supported, while hasMany relation cannot facilitate this feature without a through model.
packages repository and repository-tests are updated.

related to #2487

Signed-off-by: David Zhang <dwzhangdavid@gmail.com>

## Checklist

- [x] DCO (Developer Certificate of Origin) [signed in all commits](https://loopback.io/doc/en/contrib/code-contrib.html)
- [x] `npm test` passes on your machine
- [x] New tests added or existing tests modified to cover all changes
- [x] Code conforms with the [style guide](http://loopback.io/doc/en/contrib/style-guide.html)
- [x] API Documentation in code was updated
- [x] Documentation in [/docs/site](../tree/master/docs/site) was updated
- [ ] Affected artifact templates in `packages/cli` were updated
- [ ] Affected example projects in `examples/*` were updated

👉 [Check out how to submit a PR](https://loopback.io/doc/en/lb4/submitting_a_pr.html) 👈

### defining polymorphic models
define the polymorphic models
```
class Post extends Entity {
id;
}

class Letter extends Post {
}

class Parcel extends Post {
}
```
define the polymorphic property
```
class Delivery extends Entity {
// use belonging can be any subclass of Post, i.e. Letter or Parcel
@hasOne(() => Post, {polymorphic: true})
  post: Post;

postType: string;
}
```
or a customized discriminator
```
class Delivery extends Entity {
// use belonging can be any subclass of Post, i.e. Letter or Parcel
@hasOne(() => Post, {polymorphic: {discriminator: "post_type"}})
  post: Post;

post_type: string;
}
```
### setting up repository
```
export class DeliveryRepository extends DefaultCrudRepository {
  public readonly post: HasOneRepositoryFactory<Post, typeof Delivery.prototype.id>;

  constructor(
    dataSource: juggler.DataSource,
    @repository.getter('LetterRepository')
    protected letterRepositoryGetter: Getter<EntityCrudRepository<Post, typeof Post.prototype.id, PostRelations>>,
    @repository.getter('ParcelRepository')
    protected parcelRepositoryGetter: Getter<EntityCrudRepository<Post, typeof Post.prototype.id, PostRelations>>,
  ) {
    super(Delivery, dataSource);
    this.post = this.createHasOneRepositoryFactoryFor(
      'post',
      // use a dictionary of repoGetters instead of a single repoGetter instance
      {Letter: letterRepositoryGetter, Parcel: parcelRepositoryGetter},
    );
    this.registerInclusionResolver('post', this.post.inclusionResolver);
  }
}
```
### Query with inclusions
```
const result = await deliveryRepo.findById(delivery1.id, {
  include: ['post'],
});
```
### Creation with polymorphic relations
```
deliveryRepo
.post(deliveryId)
.create(letter1Data, {polymorphicType: "Letter"});
```

### CODE REVIEW

**Review by achrinza — APPROVED** [review]
Really sorry that I couldn't review this earlier. I've checked the PR and there doesn't seem to be any issue with the code and tests. I've pinged the other maintainers to get their input.

**Comment by OnTheThirdDay** [comment]
Cool. If you find the current implementation is fine, I will update the doc/cli/example, and add the feature to support polymorphic types search in inclusion filter.

**Comment by OnTheThirdDay** [comment]
Previously(the initial commit to support polymorphic types), if an included relation points to a polymorphic type, then the repository only see it as the base type, so that deeper inclusion is not supported. This commit tries to support a fine-grained inclusion filter by specifying `targetType` of the polymorphic relation. e.g. we have `Person` and `Student` as its subclass, then we want to include `studentId` field for `Student` polymorphic type when including a `person` relation, we can do
```
include: [
  {
    relation: 'person',
    scope: {
      include: [
        {
          relation: 'studentId',
          targetType: 'Student',
        },
      ],
    },
  },
]
```

**Comment by OnTheThirdDay** [comment]
I am half way updating the cli, but I am recently too busy to finish it soon. How about I update the docs first after you review the code, and later add the support for cli?

**Comment by achrinza** [comment]
Sorry for the delay on getting back on this (Most maintainers are catching up with their own December backlog); I've pinged them and there wasn't any objection to this PR so we should be good to merge.

**Comment by OnTheThirdDay** [comment]
Cool. I will update the doc later. btw, should it be a single section for polymorphic types or mention this feature in individual relations?

**Comment by achrinza** [comment]
It should be it's own page, similar to the other relations.

**Comment by roberttaylortech** [comment]
@OnTheThirdDay Hoping to find some time over the next week or so to review this whole feature and the PR and see how it has come together. :)

Also... could you and I get a little chat going via email? rtaylor.info (at) gmail Some cool things to chat about as well.

**Comment by lygstate** [comment]
can this be merged now?

**Comment by OnTheThirdDay** [comment]
> can this be merged now?

Code-wise I think so, but haven't finish the documentation...

**Comment by OnTheThirdDay** [comment]
I have added the documentation. I feel this feature is very straightforward. @achrinza Please have a check.

**Comment by achrinza** [comment]
The docs LGTM :+1:

Assuming there's no test failures, the last step is to squash it into a single commit.

If you'd like to continue the work in adding CLI scaffolding support, we can start a new PR after this is merged.

**Comment by OnTheThirdDay** [comment]
@achrinza Done :)

**Comment by achrinza** [comment]
Thanks for your contributions, @OnTheThirdDay! Your PR has been merged 🎉

**Comment by achrinza** [comment]
Hi @mikeevstropov, the semantics of the `@hasOne` decorator has not changed - Even with polymorphic relations, the foreign key is not at the source. The difference is the addition of a discriminator at the source model

Some examples can be found in the docs: https://loopback.io/doc/en/lb4/Polymorphic-relation.html

**Comment by mikeevstropov** [comment]
@achrinza Hi! Thank you for reply. As you said, example shows hasOne decorator in the same model with discriminator, and it looks weird. Just to be clear, previously, hasOne/hasMany may be applied to the model **B** only if the model **A** has foreignKey and discriminator defined by BelongsTo, because hasOne/hasMany is used to resolve relation in the opposite direction only.

**Comment by OnTheThirdDay** [comment]
Hi, in LB4 the polymorphism is extended to the target model. In LB3, A belongsTo B, B hasOne A, the B is a polymorphic type(imageable either Product or Employee), while in LB4, either A or B can be polymorphic.

**Comment by mikeevstropov** [comment]
@OnTheThirdDay Hi! If so, do we have the diffirence between belongsTo and hasOne? If no, how to resolve relation in the opposite direction (get Delivery from Letter or Parcel)? In lb3 you are using hasOne/hasMany for that.

**Comment by mikeevstropov** [comment]
Okay, got it. New feature of hasOne adds discriminator to the source and fk to the target and this approach is not matched to relations design of LB2/3. Based on this, the relation between hasOne/hasMany <-> belongsTo is violated. Also we can't migrate to LB4 with polymorphic feature without change the data, because previously the discriminator and fk was at the same model. Guys, do the LB4 trying to save backward compatible with the previous version?

**Comment by OnTheThirdDay** [comment]
Hi, I am not sure if I am getting it right -- do you mean in LB2/3 A belongsTo B(superclass of C and D), discriminator is on A? I think this is identical to the LB4 implementation. For fk, I think we still have keyTo on A.

**Comment by mikeevstropov** [comment]
BelongsTo adds a foreignKey to the source.
HasOne/hasMany looks at the foreignKey in target.

Polymorphic belongsTo adds a foreignKey and discriminator to the source.
Polymorphic hasOne/hasMany **just looks at the foreignKey and discriminator in target**.

HasOne/hasMany is used on the “other side” of a belongsTo relation.
**HasOne/hasMany does not extend the source or target by any extra fields.**

Polymorphic belongsTo adds `imageableId` and `imageableType` to the source - Picture.
```
{
  "name": "Picture",
  "base": "PersistedModel",
  ...
  "relations": {
    "imageable": {
      "type": "belongsTo",
      "polymorphic": true
    }
  },
...
```
Picture with related Employee.
```json
{
  "id": "1-picture",
  "imageableId": "1-employee",
  "imageableType": "Employee"
  "imageable": {
    "id": "1-employee",
    "name": "Miroslav Bajtoš"
  }
}
```
Polymorphic hasOne/hasMany looks at the `imageableId` and `imageableType` fields stored in target model. The hasMany definition is almost same as hasOne, the only difference is type should be hasMany (to get list of pictures).
```
{
  "name": "Employee",
  "base": "PersistedModel",
  ...
  "relations": {
    "picture": {
      "type": "hasOne",
      "model": "Picture",
      "polymorphic": "imageable"
    }
  }
...
```
Employee with related Picture.
```json
{
  "id": "1-employee",
  "name": "Miroslav Bajtoš"
  "picture": {
    "id": "1-picture",
    "imageableId": "1-employee",
    "imageableType": "Employee"
  }
},
```
HasOne relation is a degenerate case of a hasMany. Does the new inplementation of polymorphic hasOne resolves the target by foreignKey and discriminator stored in target? Do we have polymorphic belongsTo to store a foreignKey and discriminator at source?

**Comment by OnTheThirdDay** [comment]
I think we are not adding new fields for hasOne and hasMany when they are only served as the opposite direction of a polymorphic belongsTo.

**Comment by mikeevstropov** [comment]
@OnTheThirdDay Okay, does the new inplementation of polymorphic hasOne resolves the target by foreignKey and discriminator stored in target?

**Comment by OnTheThirdDay** [comment]
I think currently the discriminator is not checked, which may cause some problem. I will write a fix.

**Comment by mikeevstropov** [comment]
> I think we are not adding new fields for hasOne and hasMany when they are only served as the opposite direction of a polymorphic belongsTo.

Also, polymorphic hasOne/hasMany were intended to resolve 'other side' of polymorphic belongsTo and had no behaviors other than to resolve opposite direction.

**Comment by OnTheThirdDay** [comment]
Sorry that I am too busy recently, so still not being able to find time to make a fix to this issue. Hopefully I can find some time in the coming few months. Also happy to assist if anyone's willing to pick it up.

**Pull request merged**
